### ISSUE TITLE
Find expansion matches with contractions

### ISSUE THREAD

**Issue by PinkDuck** [osm-search/Nominatim#2618]
e.g. a search for "saint benedicts street, norwich" returns no street, but a few addresses.

The street has been named "St Benedicts Street" in OSM (something I disagree with), where the "St" prefix is a contraction without the dot (British style) for "Saint". Effectively an alias of, without short_name tag as the author believes it to be the official naming.

Shouldn’t Nominatim use some kind of stemming in its search textual matching?

The reverse situation works as expected, e.g. search for "st stephens square, norwich" or "saint stephens square, norwich" matches residential with street name "Saint Stephens Square".

**Comment by lonvia** [comment]
Nope, it's not a bug, it's a feature. It was a conscious choice not to do expand shortened names anymore, given that the OSM community has now widely accepted and applied the rule not to abbreviate unless the unabbreviated form is never used.

The British community has made a conscious choice, too, when claiming the exception from the rule. I respect that and you probably should respect that, too. Don't start tagging wars to make the data compliant to whatever address lists you want to geocode. According to [this help answer](https://help.openstreetmap.org/questions/19609/saint-or-st-is-there-an-official-osm-policy) their reasoning is that 'St' is the only legal spelling and the only one used. And that pretty much gets us in the realm of 'unless the unabbreviated form is never used'. So that's fine with me and in line with how Nominatim behaves.

As I said, you can set up your won Nominatim instance that does the abbreviation expansion as well. I'll look into creating a sanitizer which makes this relatively easy.

**Comment by PinkDuck** [comment]
In the tagging example I gave the legal name for that particular road is actually the expanded form, which shows the over-zealous changes made by the unhelpful original edit. I note from the import style list that official_name key is supported. I still think that if you are providing a search service that it would be a better outcome in situations such as this to continue to support the expanded search term. I can only point this out and ask, or create a unique instance of my own to mitigate this particular issue. I just feel this is the sort of thing there should be a global solution in the first instance, and that likely requires a design/implementation change. Perhaps with the OpenStreetMap installation of Nominatim itself, since a badly working search index won't help adoption.

**Comment by PinkDuck** [comment]
Am I right in understanding that existing build process adds additional contracted forms to the analysis set for matching?

So if a street name is prefixed with "Saint ", both "Saint" and "St" forms exist.
If a street name is prefixed with "St ", only "St " exists in the database to match on.

So when the user searches for "Saint " (or any known expanded prefix form), couldn't a second contracted term be added into just the user search tokens to ensure result match? While this makes the sanitising uneven between build and search, it might solve the results outcome in a general way to counter the choice to not expand shortened names. Perhaps that is what you were proposing.

**Comment by lonvia** [comment]
Adding expansion rules has become very easy via the new [derive-names sanitizer](https://github.com/osm-search/Nominatim/pull/4097). For example, to expand all leading 'St' in names in the UK just add the following sanitizer rule:

```
- step: derive-names
  type: name
  filter-country: gb
  name-pattern: 'St (.*)'
  variants: 'Saint \1'
```
I have no plans to make this part of the standard configuration for Nominatim and also no plans to add this for nominatim.openstreetmap.org. As the British OSM community claims that St is never ever written as Saint, so there is no need to do so.

**Comment by PinkDuck** [comment]
Sigh, that's basically WontFix, but thanks for sanitiser rule idea.

Just wish that searching for 'saint benedicts street, norwich” would return a result by default.

[Saint Hill Road](https://www.google.co.uk/maps/place/Saint+Hill+Rd,+East+Grinstead/@51.1058103,-0.0237756,3a,18y,231.41h,82.12t/data=!3m8!1e1!3m6!1sTwGCnckpG9fR5FNZ1bgUEw!2e0!5s20210701T000000!6shttps:%2F%2Fstreetviewpixels-pa.googleapis.com%2Fv1%2Fthumbnail%3Fcb_client%3Dmaps_sv.tactile%26w%3D900%26h%3D600%26pitch%3D7.881196856910748%26panoid%3DTwGCnckpG9fR5FNZ1bgUEw%26yaw%3D231.40833299938473!7i16384!8i8192!4m6!3m5!1s0x4875f66c47445461:0x6202866df3c05596!8m2!3d51.1089208!4d-0.0289487!16s%2Fg%2F1tjthrhg?hl=en&entry=ttu&g_ep=EgoyMDI2MDUxMy4wIKXMDSoASAFQAw%3D%3D)
